mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageIdMapper.java

```diff
@@ -0,0 +1,124 @@
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+package org.apache.james.mailbox.cassandra.mail;
+
+import static org.apache.james.mailbox.cassandra.table.CassandraMessageIds.MESSAGE_ID;
+
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import org.apache.james.backends.cassandra.utils.CassandraUtils;
+import org.apache.james.mailbox.cassandra.CassandraId;
+import org.apache.james.mailbox.cassandra.CassandraMessageId;
+import org.apache.james.mailbox.exception.MailboxException;
+import org.apache.james.mailbox.exception.MailboxNotFoundException;
+import org.apache.james.mailbox.model.ComposedMessageId;
+import org.apache.james.mailbox.model.MailboxId;
+import org.apache.james.mailbox.model.MessageId;
+import org.apache.james.mailbox.store.mail.AttachmentMapper;
+import org.apache.james.mailbox.store.mail.MailboxMapper;
+import org.apache.james.mailbox.store.mail.MessageIdMapper;
+import org.apache.james.mailbox.store.mail.MessageMapper.FetchType;
+import org.apache.james.mailbox.store.mail.model.MailboxMessage;
+import org.apache.james.mailbox.store.mail.model.Message;
+
+import com.datastax.driver.core.Row;
+import com.github.steveash.guavate.Guavate;
+import com.google.common.base.Throwables;
+
+public class CassandraMessageIdMapper implements MessageIdMapper {
+
+    private final AttachmentMapper attachmentMapper;
+    private final MailboxMapper mailboxMapper;
+    private final CassandraImapUidDAO imapUidDAO;
+    private final CassandraMessageIdDAO messageIdDAO;
+    private final CassandraMessageDAO messageDAO;
+
+    public CassandraMessageIdMapper(AttachmentMapper attachmentMapper, MailboxMapper mailboxMapper,
+            CassandraImapUidDAO imapUidDAO, CassandraMessageIdDAO messageIdDAO, CassandraMessageDAO messageDAO) {
+        this.attachmentMapper = attachmentMapper;
+        this.mailboxMapper = mailboxMapper;
+        this.imapUidDAO = imapUidDAO;
+        this.messageIdDAO = messageIdDAO;
+        this.messageDAO = messageDAO;
+    }
+
+    @Override
+    public List<Message> find(List<MessageId> messageIds, FetchType fetchType) {
+        List<CassandraMessageId> cassandraMessageIds = messageIds.stream()
+            .map(id -> (CassandraMessageId) id)
+            .collect(Collectors.toList());
+        return CassandraUtils.convertToStream(messageDAO.retrieveMessages(cassandraMessageIds, fetchType, Optional.empty()).join())
+                .map(row -> message(row, fetchType))
+                .sorted(Comparator.comparing(MailboxMessage::getUid))
```

Comment:
Do we need order at all ? If so we need messages grouped by mailbox ID if you want to sort by UID, as UID is mailbox specific. Sorting only by UID is IMO a non sens...